File: clients/lumify-sdk/src/webhook-signature.ts
// Verifies the `Lumify-Signature` header webhook deliveries carry (see
// core/webhooks/delivery.py's `_sign`):
//
//   Lumify-Signature: t=<unix_ts>,v1=<hex hmac_sha256(signing_secret, `${ts}.${body}`)>
//
// Uses Web Crypto (`crypto.subtle`), available unflagged in Node 18+ and every
// modern browser/edge runtime — no `node:crypto` import, so this stays portable
// and dependency-free.

export interface VerifyWebhookOptions {
  /** Max age (seconds) a signature is accepted, to reject replayed deliveries. Default 300 (5 min). */
  toleranceSeconds?: number;
  /** Injectable clock for tests. */
  now?: () => number;
}

export class WebhookSignatureError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "WebhookSignatureError";
  }
}

function parseSignatureHeader(header: string): { timestamp: number; signature: string } {
  const parts = new Map<string, string>();
  for (const segment of header.split(",")) {
    const [key, value] = segment.split("=", 2);
    if (key && value !== undefined) parts.set(key.trim(), value.trim());
  }
  const t = parts.get("t");
  const v1 = parts.get("v1");
  if (!t || !v1) {
    throw new WebhookSignatureError(
      "Malformed Lumify-Signature header — expected 't=<unix_ts>,v1=<hex_hmac>'."
    );
  }
  const timestamp = Number(t);
  if (!Number.isFinite(timestamp)) {
    throw new WebhookSignatureError("Malformed Lumify-Signature header — 't' is not a number.");
  }
  return { timestamp, signature: v1 };
}

async function hmacSha256Hex(secret: string, payload: string): Promise<string> {
  const enc = new TextEncoder();
  const key = await crypto.subtle.importKey(
    "raw",
    enc.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"]
  );
  const sig = await crypto.subtle.sign("HMAC", key, enc.encode(payload));
  return Array.from(new Uint8Array(sig))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

/** Constant-time string comparison (no early exit on the first mismatched byte). */
function timingSafeEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return diff === 0;
}

/**
 * Verify a webhook delivery's `Lumify-Signature` header against the raw
 * request body and your subscription's `signing_secret`. Throws
 * {@link WebhookSignatureError} on any failure (bad format, mismatch, stale
 * timestamp) — treat that as "reject the delivery with a 4xx", not a crash.
 *
 * `body` must be the *raw, unparsed* request bytes/string — re-serializing
 * parsed JSON can reorder keys or change whitespace and break the signature.
 *
 * @example
 * app.post("/webhooks/lumify", express.raw({ type: "*\/*" }), async (req, res) => {
 *   try {
 *     await verifyWebhook(process.env.LUMIFY_WEBHOOK_SECRET!, req.header("Lumify-Signature")!, req.body.toString("utf8"));
 *   } catch {
 *     return res.status(400).end();
 *   }
 *   res.status(200).end();
 * });
 */
export async function verifyWebhook(
  signingSecret: string,
  signatureHeader: string,
  rawBody: string,
  options: VerifyWebhookOptions = {}
): Promise<void> {
  const { timestamp, signature } = parseSignatureHeader(signatureHeader);

  const tolerance = options.toleranceSeconds ?? 300;
  const now = (options.now ?? (() => Date.now() / 1000))();
  if (Math.abs(now - timestamp) > tolerance) {
    throw new WebhookSignatureError(
      `Signature timestamp is ${Math.round(Math.abs(now - timestamp))}s old, outside the ${tolerance}s tolerance (possible replay).`
    );
  }

  const expected = await hmacSha256Hex(signingSecret, `${timestamp}.${rawBody}`);
  if (!timingSafeEqual(expected, signature)) {
    throw new WebhookSignatureError("Signature mismatch — payload may have been tampered with, or the signing secret is wrong.");
  }
}

File: clients/lumify-sdk-python/lumify/webhook_signature.py
"""
Verifies the ``Lumify-Signature`` header webhook deliveries carry (see
core/webhooks/delivery.py's ``_sign``):

    Lumify-Signature: t=<unix_ts>,v1=<hex hmac_sha256(signing_secret, `${ts}.${body}`)>

Uses stdlib ``hmac``/``hashlib`` — no dependencies. Mirrors the TS SDK's
``verifyWebhook`` (which uses Web Crypto).
"""

from __future__ import annotations

import hashlib
import hmac
import time
from typing import Callable, Optional, Tuple


class WebhookSignatureError(Exception):
    """Raised when a webhook signature is malformed, mismatched, or stale."""


def _parse_signature_header(header: str) -> Tuple[int, str]:
    parts = {}
    for segment in header.split(","):
        key, sep, value = segment.partition("=")
        if sep and value != "":
            parts[key.strip()] = value.strip()
    t = parts.get("t")
    v1 = parts.get("v1")
    if not t or not v1:
        raise WebhookSignatureError(
            "Malformed Lumify-Signature header — expected 't=<unix_ts>,v1=<hex_hmac>'."
        )
    try:
        timestamp = int(float(t))
    except (TypeError, ValueError):
        raise WebhookSignatureError(
            "Malformed Lumify-Signature header — 't' is not a number."
        )
    return timestamp, v1


def verify_webhook(
    signing_secret: str,
    signature_header: str,
    raw_body: str,
    *,
    tolerance_seconds: int = 300,
    now: Optional[Callable[[], float]] = None,
) -> None:
    """Verify a delivery's ``Lumify-Signature`` header against the raw request
    body and your subscription's ``signing_secret``.

    Raises :class:`WebhookSignatureError` on any failure (bad format, mismatch,
    stale timestamp) — treat that as "reject the delivery with a 4xx", not a
    crash.

    ``raw_body`` must be the *raw, unparsed* request body — re-serializing
    parsed JSON can reorder keys or change whitespace and break the signature.
    """
    timestamp, signature = _parse_signature_header(signature_header)

    now_val = (now or time.time)()
    if abs(now_val - timestamp) > tolerance_seconds:
        raise WebhookSignatureError(
            "Signature timestamp is %ds old, outside the %ds tolerance "
            "(possible replay)." % (round(abs(now_val - timestamp)), tolerance_seconds)
        )

    expected = hmac.new(
        signing_secret.encode("utf-8"),
        ("%d.%s" % (timestamp, raw_body)).encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    if not hmac.compare_digest(expected, signature):
        raise WebhookSignatureError(
            "Signature mismatch — payload may have been tampered with, or the "
            "signing secret is wrong."
        )
